lib/src/Base/Optim/Bonmin.cxx

```diff
@@ -0,0 +1,312 @@
+/**
+ *  @brief BonminProblem implements the interface between OT and Bonmin as
+ * for optimization problems. It is derived from Bonmin::TMINLP to ensure
+ * compatibility with Bonmin algorithms.
+ *
+ *  Copyright 2005-2019 Airbus-EDF-IMACS-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/Bonmin.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/OSS.hxx"
+#include "openturns/SymbolicFunction.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#ifdef OPENTURNS_HAVE_BONMIN
+  #include "openturns/BonminProblem.hxx"
+  #include "BonBonminSetup.hpp"
+  #include "BonCbc.hpp"
+  using namespace Bonmin;
+  using namespace Ipopt;
+#endif
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(Bonmin)
+
+static const Factory<Bonmin> Factory_Bonmin;
+
+/** Constructors */
+Bonmin::Bonmin(const String & algoName)
+  : OptimizationAlgorithmImplementation()
+  , algoName_(algoName)
+  {
+    // Nothing to do
+  }   
+
+Bonmin::Bonmin( OptimizationProblem & problem,
+                const String & algoName)
+  : OptimizationAlgorithmImplementation(problem)
+  , algoName_()
+  {
+    setAlgorithmName(algoName);
+  }
+
+/* Virtual constructor */
+Bonmin * Bonmin::clone() const
+{
+  return new Bonmin(*this);
+}
+
+  
+  
+  /** Bonmin static methods */
+  Bool Bonmin::IsAvailable()
+  {
+    #ifdef OPENTURNS_HAVE_BONMIN
+      return true;
+    #else
+      return false;
+    #endif
+  }
+  
+  Description Bonmin::GetAlgorithmNames()
+  {
+    Description algoNames(5);
+    algoNames[0] = "B-BB";
+    algoNames[1] = "B-OA";
+    algoNames[2] = "B-QG";
+    algoNames[3] = "B-Hyb";
+    algoNames[4] = "B-iFP";
+    
+    return algoNames;
+  }
+
+  /** Accessors to Bonmin attributes */
+    // Algorithm name
+  void Bonmin::setAlgorithmName(const String & algoName)
+  {
+    // Check algoName
+    if (!GetAlgorithmNames().contains(algoName))
+    {
+      if (algoName == "B-Ecp")
+      {
+        throw InternalException(HERE) << "B-Ecp solver is unavailable.";
+      }
+      else
+      {
+        throw InvalidArgumentException(HERE) << "Unknown solver " << algoName;
+      }
+    }
+      
+    algoName_ = algoName;
+  }
+  
+  String Bonmin::getAlgorithmName() const
+  {
+    return algoName_;
+  }
+
```

Comment:
Strange whitespace

Reply:
I'm not sure what you meant... I removed some blank lines and useless brackets to get a more condensed code.
Did you have something else in mind?

Reply:
just that you got a strange indentation (and not in all the `cxx` file)
Maybe it is just linked to your editor configuration ? Btw it is not important, Julien will apply the `reindent` script before the release